File: checkout/java/src/main/java/org/pivotal/katas/Checkout.java
package org.pivotal.katas;

import java.util.Map;

public class Checkout {
  private final Map<String, Pricing> rules;

  public Checkout(Map<String, Pricing> rules) {
    this.rules = rules;
  }

  public void scan(String item) {

  }

  public int total() {
    return 0;
  }

  public static class Discount {
    private final int quantity;
    private final int price;

    public Discount(int quantity, int price) {
      this.quantity = quantity;
      this.price = price;
    }
  }

  public static class Pricing {
    private final int price;
    private final Discount discount;

    public Pricing(int price) {
      this(price, null);
    }

    public Pricing(int price, Discount discount) {
      this.price = price;
      this.discount = discount;
    }
  }
}

File: checkout/ruby/lib/checkout.rb
class Checkout
  def initialize(rules)
    @rules = rules
    @items = []
  end

  def scan(item)
    @items << item
  end

  def total
    @grouped = @items.reduce(Hash.new(0)) do |hash, item|
      hash[item] += 1
      hash
    end

    @grouped.reduce(0) do |total, (item, quantity)|
      rule = @rules[item]
      price, discount = rule
      if discount
        dquantity, dprice = discount
        total += dprice * (quantity / dquantity).floor
        quantity = quantity % dquantity
      end
      total + price * quantity
    end
  end
end